transport/src/test/java/io/netty/channel/embedded/EmbeddedChannelTest.java

```diff
@@ -380,6 +384,161 @@ public void run() {
         assertNull(channel.readOutbound());
     }
 
+    @Test
+    public void testFlushInbound() throws InterruptedException {
+        final CountDownLatch latch = new CountDownLatch(1);
+        EmbeddedChannel channel = new EmbeddedChannel(new ChannelInboundHandlerAdapter() {
+            @Override
+            public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
+              latch.countDown();
+            }
+        });
+
+        channel.flushInbound();
+
+        if (!latch.await(1L, TimeUnit.SECONDS)) {
+            fail("Nobody called #channelReadComplete() in time.");
+        }
+    }
+
+    @Test
+    public void testWriteOneInbound() throws InterruptedException {
+      final CountDownLatch latch = new CountDownLatch(1);
+      final AtomicInteger flushCount = new AtomicInteger(0);
+
+      EmbeddedChannel channel = new EmbeddedChannel(new ChannelInboundHandlerAdapter() {
+          @Override
+          public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+              latch.countDown();
+          }
+
+          @Override
+          public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
+              flushCount.incrementAndGet();
+          }
+      });
+
+      channel.writeOneInbound("Hello, Netty!");
+
+      if (!latch.await(1L, TimeUnit.SECONDS)) {
+          fail("Nobody called #channelRead() in time.");
+      }
+
+      channel.close().syncUninterruptibly();
+
+      // There was no #flushInbound() call so nobody should have called
+      // #channelReadComplete()
+      assertEquals(flushCount.get(), 0);
+    }
+
+    @Test
+    public void testFlushOutbound() throws InterruptedException {
+        final CountDownLatch latch = new CountDownLatch(1);
+        EmbeddedChannel channel = new EmbeddedChannel(new ChannelOutboundHandlerAdapter() {
+            @Override
+            public void flush(ChannelHandlerContext ctx) throws Exception {
+                latch.countDown();
+            }
+        });
+
+        channel.flushOutbound();
+
+        if (!latch.await(1L, TimeUnit.SECONDS)) {
+            fail("Nobody called #flush() in time.");
+        }
+    }
+
+    @Test
+    public void testWriteOneOutbound() throws InterruptedException {
+        final CountDownLatch latch = new CountDownLatch(1);
+        final AtomicInteger flushCount = new AtomicInteger(0);
+
+        EmbeddedChannel channel = new EmbeddedChannel(new ChannelOutboundHandlerAdapter() {
+            @Override
+            public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+                latch.countDown();
```

Comment:
would it impact the test if this came after the `ctx.write()` call (maybe somehow this would call flush)?